File: Examples/perl/basics.pl
use strict;
use warnings;

sub fib {
    my ($n) = @_;
    return $n if $n < 2;
    return fib($n - 1) + fib($n - 2);
}

sub greet {
    my ($name, $greeting) = @_;
    return "$greeting, $name!";
}

my @fibs;
for (my $i = 0; $i < 10; $i++) {
    push @fibs, fib($i);
}
print "fib: ", join(", ", @fibs), "\n";

print greet("World", "Hello"), "\n";

my @nums = (5, 3, 9, 1, 7);
my @sorted = sort { $a <=> $b } @nums;
print "sorted: @sorted\n";
print "count: ", scalar(@nums), "\n";

my @doubled = map { $_ * 2 } @nums;
print "doubled: @doubled\n";

my @big = grep { $_ > 3 } @nums;
print "big: @big\n";

my $total = 0;
foreach my $n (@nums) {
    $total += $n;
}
print "total: $total\n";

my %ages = ("alice", 30, "bob", 25);
$ages{"carol"} = 35;
foreach my $name (sort keys %ages) {
    print "$name is $ages{$name}\n";
}

my $text = "Hello, Perl World";
print "upper: ", uc($text), "\n";
print "length: ", length($text), "\n";
print "words: ", join("|", split(/,?\s+/, $text)), "\n";

if ($text =~ /Perl/) {
    print "matched Perl\n";
}

my $copy = $text;
$copy =~ s/World/Hackers/;
print "replaced: $copy\n";

my $i = 0;
while ($i < 3) {
    print "i=$i\n";
    $i++;
}

unless (0) {
    print "unless works\n";
}

printf("pi is about %.3f\n", 22 / 7);
print "17 % 5 = ", 17 % 5, "\n";
print "2 ** 10 = ", 2 ** 10, "\n";
print "concat: " . "a" . "b" . "\n";
print "repeat: ", "-" x 10, "\n";

File: Examples/r/basics.R
# R の基本
fib <- function(n) {
  if (n < 2) {
    return(n)
  }
  fib(n - 1) + fib(n - 2)
}

fibs <- sapply(0:9, fib)
cat("fib:", fibs, "\n")

x <- c(5, 3, 9, 1, 7)
cat("sum:", sum(x), "\n")
cat("mean:", mean(x), "\n")
cat("sorted:", sort(x), "\n")
cat("rev:", rev(x), "\n")
cat("length:", length(x), "\n")
cat("max/min:", max(x), min(x), "\n")

y <- x * 2
cat("doubled:", y, "\n")
cat("plus one:", x + 1, "\n")
cat("big:", x[x > 3], "\n")

squares <- sapply(1:5, function(v) v^2)
cat("squares:", squares, "\n")

total <- 0
for (i in 1:10) {
  total <- total + i
}
cat("total:", total, "\n")

i <- 0
while (i < 3) {
  cat("i =", i, "\n")
  i <- i + 1
}

greet <- function(name, greeting = "Hello") {
  paste0(greeting, ", ", name, "!")
}
cat(greet("World"), "\n")
cat(greet("R", greeting = "Hi"), "\n")

person <- list(name = "Alice", age = 30)
cat("name:", person$name, "age:", person$age, "\n")

s <- "Hello, R Language"
cat("upper:", toupper(s), "\n")
cat("nchar:", nchar(s), "\n")
cat("substr:", substr(s, 1, 5), "\n")

cat("17 %% 5 =", 17 %% 5, "\n")
cat("17 %/% 5 =", 17 %/% 5, "\n")
cat("3 %in% x:", 3 %in% x, "\n")

print(x)
print(sort(x))
